lib/AST/ASTScope.cpp

```diff
@@ -39,6 +40,103 @@ using namespace ast_scope;
 
 #pragma mark ASTScope
 
+class LoggingASTScopeDeclConsumer
+    : public namelookup::AbstractASTScopeDeclConsumer {
+private:
+  namelookup::AbstractASTScopeDeclConsumer *originalConsumer;
+
+public:
+  mutable SmallVector<BridgedConsumedLookupResult> recordedElements;
+
+  LoggingASTScopeDeclConsumer(
+      namelookup::AbstractASTScopeDeclConsumer *consumer)
+      : originalConsumer(consumer) {}
+
+  ~LoggingASTScopeDeclConsumer() = default;
+
+  /// Called for every ValueDecl visible from the lookup.
+  ///
+  /// Takes an array in order to batch the consumption before setting
+  /// IndexOfFirstOuterResult when necessary.
+  ///
+  /// Additionally, each name is logged to `recordedElements` and
+  /// can be later used in validation of `SwiftLexicalLookup` result.
+  ///
+  /// \param baseDC either a type context or the local context of a
+  /// `self` parameter declaration. See LookupResult for a discussion
+  /// of type -vs- instance lookup results.
+  ///
+  /// \return true if the lookup should be stopped at this point.
+  bool consume(ArrayRef<ValueDecl *> values,
+               NullablePtr<DeclContext> baseDC = nullptr) override {
+    bool result = originalConsumer->consume(values, baseDC);
+
+    for (auto value : values) {
+      if (auto sourceLoc = value->getLoc()) {
+        recordedElements.push_back(BridgedConsumedLookupResult(
+            value->getBaseIdentifier(), sourceLoc, result));
+      } else {
+        // If sourceLoc is unavailable, use location of it's parent.
+        recordedElements.push_back(BridgedConsumedLookupResult(
+            value->getBaseIdentifier(),
+            value->getDeclContext()->getAsDecl()->getLoc(), result));
+      }
+    }
+
+    return result;
+  };
+
+  /// Look for members of a nominal type or extension scope.
+  ///
+  /// Each call is recorded in `recordedElements` with a special flag set.
+  /// It can be later used in validation of `SwiftLexicalLookup` result.
+  ///
+  /// \return true if the lookup should be stopped at this point.
+  bool lookInMembers(const DeclContext *scopeDC) const override {
+    bool result = originalConsumer->lookInMembers(scopeDC);
+
+    if (auto *extDecl = dyn_cast<ExtensionDecl>(scopeDC)) {
+      recordedElements.push_back(BridgedConsumedLookupResult(
+          Identifier(), extDecl->getExtendedTypeRepr()->getLoc(),
+          0b10 + result));
```

Comment:
Perhaps the magic constant should be named, or you could use a bitfield or something

Reply:
This is a good idea. I documented those on the Swift side, but forgot about explaining them here. I moved `0b10` to a constant `shouldLookInMembers` and also renamed `result` to `endOfLookup`.